packages/service-library/tests/test_rabbitmq_robust_rpc.py

```diff
@@ -0,0 +1,229 @@
+    rpc_client: RobustRPCClient, rpc_server: RobustRPCServer
+):
+    await rpc_server.register(add_me)
+
+    for i in range(MULTIPLE_REQUESTS_COUNT):
+        assert await rpc_client.request(add_me.__name__, x=1 + i, y=2 + i) == 3 + i * 2
+
+
+async def test_multiple_requests_parallel_same_server_and_client(
+    rpc_client: RobustRPCClient, rpc_server: RobustRPCServer
+):
+    await rpc_server.register(add_me)
+
+    expected_result: list[int] = []
+    requests: list[Awaitable] = []
+    for i in range(MULTIPLE_REQUESTS_COUNT):
+        requests.append(rpc_client.request(add_me.__name__, x=1 + i, y=2 + i))
+        expected_result.append(3 + i * 2)
+
+    assert await asyncio.gather(*requests) == expected_result
+
+
+async def test_multiple_requests_parallel_same_server_different_clients(
+    rabbit_service: RabbitSettings, rpc_server: RobustRPCServer
+):
+    await rpc_server.register(add_me)
+
+    clients: list[RobustRPCClient] = []
+    for _ in range(MULTIPLE_REQUESTS_COUNT):
+        client = RobustRPCClient(rabbit_service)
+        clients.append(client)
+
+    # worst case scenario
+    await asyncio.gather(*[c.start() for c in clients])
+
+    requests: list[Awaitable] = []
+    expected_result: list[int] = []
+    for i in range(MULTIPLE_REQUESTS_COUNT):
+        client = clients[i]
+        requests.append(client.request(add_me.__name__, x=1 + i, y=2 + i))
+        expected_result.append(3 + i * 2)
+
+    assert await asyncio.gather(*requests) == expected_result
+
+    # worst case scenario
+    await asyncio.gather(*[c.stop() for c in clients])
+
+
+async def test_raise_error_if_not_started(rabbit_service: RabbitSettings):
+    client = RobustRPCClient(rabbit_settings=rabbit_service)
+    with pytest.raises(NotStartedError):
+        await client.request(add_me, x=1, y=2)
+
+    # expect not to raise error
+    await client.stop()
+
+    server = RobustRPCServer(rabbit_settings=rabbit_service)
+    with pytest.raises(NotStartedError):
+        await server.register(add_me)
+
+    # expect not to raise error
+    await server.stop()
+
+
+async def _assert_event_not_registered(rpc_client: RobustRPCClient):
+    with pytest.raises(RemoteMethodNotRegisteredError) as exec_info:
+        assert await rpc_client.request(add_me.__name__, x=1, y=3) == 3
+    assert (
+        f"Could not find a remote method named: '{add_me.__name__}'"
+        in f"{exec_info.value}"
+    )
+
+
+async def test_server_not_started(rpc_client: RobustRPCClient):
+    await _assert_event_not_registered(rpc_client)
+
+
+async def test_server_handler_not_registered(
+    rpc_client: RobustRPCClient, rpc_server: RobustRPCServer
+):
+    await _assert_event_not_registered(rpc_client)
+
+
+async def test_request_is_missing_arguments(
+    rpc_client: RobustRPCClient, rpc_server: RobustRPCServer
+):
+    await rpc_server.register(add_me)
+
+    # missing 1 argument
+    with pytest.raises(TypeError) as exec_info:
```

Comment:
here, how does python normally handle this? I would try to mimick it.

Reply:
That is how actually python handles it, and this comes directly over the wire via aiorabbit's RPC client